### PR TITLE
Integrated warning on returning null

### PR DESCRIPTION
**Pull request by ev1stensberg**
### Things I don't know about this PR.

- **Integration with jest**
- **If we should warn at all**

***

- [ ] check if the equal comparison is correct. 

- [ ] Integrate unit test / manually test this? 

_Feel free to close this PR if this change is not appropriate( @jimfb )_ 😆

### CODE REVIEW

**Comment by StoneCypher** [comment]
Returning null is how you effectively render nothing.  The manual recommends it all over the place, and tons of codebases do this.  There is no desire to warn: this is a normal, correct behavior.

File: src/isomorphic/classic/class/ReactClass.js [hunk 1]
```diff
@@ -778,8 +778,11 @@ var ReactClass = {
         // We allow auto-mocks to proceed as if they're returning null.
         if (initialState === undefined &&
             this.getInitialState._isMockFunction) {
-          // This is probably bad practice. Consider warning here and
-          // deprecating this convenience.
+          warning(
+            typeof initialState === 'object' && initialState == null && !Array.isArray(initialState),
```
Comment (zpao) [inline src/isomorphic/classic/class/ReactClass.js]: We already know that `initialState === undefined` so this condition will never be met and we'll always warn. It's also an entirely unactionable warning as this only happens when a component is mocked.

Comment (ev1stensberg) [inline src/isomorphic/classic/class/ReactClass.js]: Hmm, okay. Closing this then 😄

**Comment by zpao** [comment]
@ev1stensberg Hey, I've noticed you've been doing a few PRs lately, which is great. However a lot of them haven't made much progress. One of the things I'd like for us to do more of is discuss changes in behavior in an issue before implementing them. This allows discussion of the idea to happen and can shape & clarify the plan before any code gets changed. For example, filing an issue declaring your goal of warning when `initialState` is `null` would have been a great place to kick off the discussion and may have lead to a clearer indication of where that change would have to be made. Keep it up, but let's make sure we focus on the right things :)

File: src/isomorphic/classic/class/ReactClass.js (hunk 1 above)
Comment (ev1stensberg) [inline src/isomorphic/classic/class/ReactClass.js]: I thought the comment was meant to warn if user returns null in initialState, (also forgot about first conditional). Is the comment meant to warn AND/OR remove the mocking function? As by now, there's only two places in the repository that uses `._isMockFunction`

**Comment by ev1stensberg** [comment]
@zpao Yeah, was thinking the same thing when I made this PR. I'm working on implementing `transferState()` and `transferProps()` in my own fork of React as per se, and noticed the comment. Just wanted to make a quick fix. For future PR's I'll do that. 

***

Also, question: Where's the roadmap for React? (Other than React future) If so, I doubt you've included all these little "issues" in the roadmap, which would be great for people that aren't necessary the best on filing quality PR's. If so, it should include what changes should include for each set of issues. What do you think? 

Also, thanks for the reminder, I know it's probably not in your best interest handling all these PR's.

**Comment by jimfb** [comment]
@ev1stensberg https://github.com/facebook/react/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+bug%22

**Comment by gaearon** [comment]
As of right now there is no specific roadmap. We might create a special repo for tracking it, but we haven’t committed to doing this yet.

In the meantime, as @jimfb notes, [good first bug](https://github.com/facebook/react/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+bug%22) is usually a good place to start. Please make sure to check that if you want to work on an issue, nobody else is already working on it, and leave comment stating you’d like to help, so nobody starts working on it right after you.

For changes like this PR, it’s absolutely fine just to raise an issue with a suggestion first. This issue tracker is a good place for this.

**Comment by ev1stensberg** [comment]
@gaearon @jimfb Yeah totally agree. Problem with issue tracker, is that the issue gets closed straight away without a good discussion that helps towards a complete solution. If its closed, it gets trashed and forgotten. Still waiting for someone to answer my closed thread on transferring state through props, but I know it won't be much of help when it's closed. 

No one is looking actively on closed threads. That's one of the reasons why I file PRs instead of issues. It has a higher response ratio than actual issues, when it should really be the other way around or at least much more balanced.

**Comment by gaearon** [comment]
>Problem with issue tracker, is that the issue gets closed straight away without a good discussion that helps towards a complete solution.

>If its closed, it gets trashed and forgotten. Still waiting for someone to answer my closed thread on transferring state through props, but I know it won't be much of help when it's closed.

Generally this doesn’t happen when we understand *why* you are trying to accomplish something, and it matches what we’d like to see in React. [This issue](https://github.com/facebook/react/issues/6114) is a good example of this at work: somebody saw a problem in React and explained it, the team agreed this is something they would like to see, and somebody else [came along to implement it](https://github.com/facebook/react/pull/6121).

However, issues like #6616 are harder to discuss because they lack the essential context:

* Why are you implementing this?
* How will this help you build your app, and will this be useful for teams at Facebook?
* Does this require API changes, and why are they justified?
* Can you show an external interest in this feature from other users of React?

When an issue fails to address these questions, keeping it open doesn’t help anyone because it is not actionable for the team or anyone else.

On the other hand, focused, easy to understand issues like #6114 that don’t affect the public API are usually a better way to start contributing.

I hope this helps! We mean well.

**Comment by ev1stensberg** [comment]
#6616 actually refer to an existing issue, and the issue was built upon trying to move state in a more freely matter. And yeah, I agree to the bulletpoints, but I think that a discussion is meant to build up a case  along the way too. For this example, it had the intent: **feedback and reiteration** And I don't really see any other way to get people to be notified about this other than the issue tracker as a informative source. Social media / stack overflow is not good streams for that kind of content, you won't guarantee to get feedback from authors.

To conclude: Feedback is gold, but in order to make a strong PR and use case, you need to have a discussion ---> of which a closed issue won't contain. Issue trackers should be used for reiteration, and I think that these bullet points are not always relevant, you need to find out as a group / community how you want to approach rather than just individuals, you'll get a better end-product.

About the last line -- Yeah you guys are awesome, so I'm taking this as constructive criticism :blue_heart:

**Comment by ev1stensberg** [comment]
Forgot to mention that you got React specific forums, but with a proposal-ish, you'll need a place to stick your code, rather than a gist. What I think we should have, is a label named "forum discussion" or "feature discussion" so people actually sees the "issue" , as by now I think we only have the discussion label on GH

**Pull request closed without merging**